### PR TITLE
HPCC-26406 Scale stub memory resource by number of instances

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: helm/hpcc/templates/eclagent.yaml [hunk 1]
```diff
@@ -163,7 +163,7 @@ spec:
 {{- if .useChildProcesses }}
 {{- include "hpcc.addResources" (dict "me" .resources) | indent 8 }}
 {{- else if not $.Values.global.privileged }}
-{{- include "hpcc.addStubResources" . | indent 8 }}
+{{- include "hpcc.addStubResources" (dict "instances" .maxActive) | indent 8 }}
```
Comment (ghalliday) [inline helm/hpcc/templates/eclagent.yaml:166]: What happens if maxActive is not defined?

Comment (jakesmith) [reply helm/hpcc/templates/eclagent.yaml:166]: good point, I meant default to 1

Comment (jakesmith) [reply helm/hpcc/templates/eclagent.yaml:166]: @ghalliday - max made maxActive properties required

File: helm/hpcc/values.schema.json [hunk 2]
```diff
@@ -799,13 +799,19 @@
...
       "properties": {
         "name": {
           "type": "string",
           "description": "The name of the eclccserver process"
         },
+        "maxActive": {
+          "type": "integer",
+          "minimum": 1,
+          "description": "Maximum number of compile jobs that can be active simultaneously",
+          "default": 4
```
Comment (ghalliday) [inline helm/hpcc/values.schema.json:813]: That does this do?

File: testing/helm/tests/env.yaml [hunk 3]
```diff
@@ -58,6 +59,7 @@ thor:
 
 eclccserver: 
 - name: eclccserver
+  maxActive: 1
```
Comment (ghalliday) [inline testing/helm/tests/env.yaml:62]: Previously this defaulted to 4.  Should discuss what the correct value is.

File: testing/helm/tests/env.yaml [hunk 4]
```diff
@@ -38,6 +38,7 @@ dfuserver:
     
 eclagent:
 - name: eclagent
+  maxActive: 1
```
Comment (ghalliday) [inline testing/helm/tests/env.yaml:41]: Previously defaulted to 100 (!?).  Should discuss what the correct value is.

**Review by ghalliday — COMMENTED** [review]
@jakesmith change looks ok, but have a few questions, which are probably worth discussing with Richard.
I would be tempted to include in 8.2.x because it must be currently causing problems, but it also is a breaking change for existing values files (since maxActive is now required), so that wouldn't be a good idea.
There is a separate issue of what the defaults should be (which is probably related to auto-scaling)

File: helm/hpcc/values.schema.json (hunk 2 above)
Comment (jakesmith) [reply helm/hpcc/values.schema.json:813]: This mirrors the default in code (in eclccserver.cpp):
```
#ifdef _CONTAINERIZED
            queryCodeSigner().initForContainer();

            bool useChildProcesses = globals->getPropInt("@useChildProcesses", false);
            unsigned maxThreads = globals->getPropInt("@maxActive", 4);
#else

```

File: testing/helm/tests/env.yaml (hunk 3 above)
Comment (jakesmith) [reply testing/helm/tests/env.yaml:62]: as it's just a schema validation file, I thought 1 (min) was fine for this test.

File: testing/helm/tests/env.yaml (hunk 4 above)
Comment (jakesmith) [reply testing/helm/tests/env.yaml:41]: as it's just a schema validation file, I thought 1 (min) was fine for this test.

Shall I change to match defaults?

Comment (ghalliday) [reply testing/helm/tests/env.yaml:41]: no, I didn't spot they were validation files - they're fine as they are.

File: helm/hpcc/values.schema.json (hunk 2 above)
Comment (ghalliday) [reply helm/hpcc/values.schema.json:813]: Does it do anything though (do some editors use it?)  Genuine question...

Comment (jakesmith) [reply helm/hpcc/values.schema.json:813]: ah, no idea.

**Comment by jakesmith** [comment]
> @jakesmith change looks ok, but have a few questions, which are probably worth discussing with Richard.
> I would be tempted to include in 8.2.x because it must be currently causing problems, but it also is a breaking change for existing values files (since maxActive is now required), so that wouldn't be a good idea.

Although it wasn't enforced in 8.2, maxActive is present in our stock values.yaml.
So yes, a custom values.yaml definition for eclagent or eclccserver that had maxActive missing, would now fail schema, but if they hadn't defined those, then it would continue to pass.

> There is a separate issue of what the defaults should be (which is probably related to auto-scaling)

Agreed, e.g. in-code 100 seems high for default for agentexec, however our stock values.yaml overrides it to 4.
As you say, slightly separate issue though.

**Comment by ghalliday** [comment]
ok.  I suspect it is ok to merge as-is to 8.2.x, but will let @richardkchapman look at it first.

**Review by richardkchapman — APPROVED** [review]

**Pull request merged**